### ISSUE TITLE
"no-unused-vars" reports only the last function parameter

### ISSUE THREAD

**Issue by julienw** [eslint/eslint#7124]
With the following code:
```javascript
const func = (req, res, next) => {};
```
only `next` is reported with the rule `no-unused-vars`.

This would not be an big issue but this actually triggers another bigger issue.

With the following code:
```javascript
const func = (req, res, _next) => {};
```
and the following configuration:
```
"no-unused-vars": [ 2, { "argsIgnorePattern": "^_" }]
```
nothing is reported at all ! Which is IMO more serious.

**Comment by kaicataldo** [comment]
Thanks for the issue report. I believe this rule is behaving as intended. It only warns on the last function parameter because you might have to define the previous parameters to be able to access it in the function body. It looks to me like `argsIgnorePattern` is also working as intended, as it then ignores any cases that match the regex (which, in your example, `_next` does). Does that make sense?

**Comment by platinumazure** [comment]
@kaicataldo I'm not sure I agree. In the first example, next is unused, so req and res should not be "protected" by the (non-)use of next. In the second example, ignoring a parameter isn't the same as using it (or at least shouldn't be), so I wouldn't want req and res to be passed over in that case either.

Both of those seem like potential bugs to me, especially the second.

**Comment by mysticatea** [comment]
Thank you for this issue.

This happens by the other option, [`args`](http://eslint.org/docs/rules/no-unused-vars#args).
The `args` option of `no-unused-vars` rule is `"after-used"` by default. This option will warn only the last argument. You can use another option `"all"`.

I'm interested in a change of the behavior.
It sounds good to me if `"after-used"` option warns the last parameter which isn't ignored.

**Comment by mysticatea** [comment]
By the way, @julienw might be interested in [no-use-ignored-vars](https://github.com/mysticatea/eslint-plugin/blob/master/docs/rules/no-use-ignored-vars.md) rule of my plugin. :smiley:

**Comment by kaicataldo** [comment]
@platinumazure Thanks for clarifying - that makes sense! I agree that an enhancement/bug fix for this would be great:
>In the second example, ignoring a parameter isn't the same as using it (or at least shouldn't be), so I wouldn't want req and res to be passed over in that case either.

I'm :+1: for @mysticatea's suggestion to alter the `"after-used"` option to warn about the last parameter that isn't ignored (rather than simply the last parameter).

**Comment by kaicataldo** [comment]
Labeled as `enhancement`, but feel free to change it to `bug` if that seems more apt.

**Comment by mysticatea** [comment]
Yeah, I think this is a bug since we can overlook unused variables.
I'm working on this.

**Comment by nzakas** [comment]
Logistics: Use both "bug" and "enhancement" labels if it's a confirmed bug and the fix would result in more warnings being reported.
